add Interest::PRIORITY and ready and ready_mut functions to AsyncFd

folkertdev opened the pull request:
## Motivation

Closes #4885

It was impossible in tokio to await `POLLPRI` events. These exceptional conditions are raised under certain circumstances on unix systems (e.g. when a new device is mounted, or when a UDP socket receives messages in its error queue).

In version [8.6.0](https://github.com/tokio-rs/mio/blob/master/CHANGELOG.md#086), `mio` adds `Interest::PRIORITY`, (for the linux and android operating systems).


## Solution

- bumps Tokio's `mio` dependency to version `8.6.0`
- exposes (publicly) `Interest::PRIORITY`
- adds `priority` and `priority_mut` functions to `AsyncFd`

## Next steps

- I decided not to expose this functionality for other socket-like types for now. If you're working with priority events, you're doing something pretty low-level and likely are using `AsyncFd` already.
- I'm not sure how/if this should be tested in tokio itself (or if `mio` is trusted)
- I'm not sure what the policy is on bumping dependencies (and specifically `mio`). Everything looks fine, but I may be missing something.

folkertdev commented:
I'm seeing a weird CI error here, that seems unrelated to the changes that are made. My other PR for which CI ran recently has the same problem

- https://github.com/tokio-rs/tokio/actions/runs/4478017069/jobs/7870270813?pr=5566
- https://github.com/tokio-rs/tokio/actions/runs/4477857160/jobs/7869917563?pr=5542

So I'm assuming this is because I rebased on current master?

<details>
  <summary>CI Output</summary>
  
     Running tests/time_delay_queue.rs (target/wasm32-wasi/debug/deps/time_delay_queue-f0c21ef322a44873.wasm)

running 30 tests
test compact_change_deadline ... ok
test compact_expire_empty ... ok
test compact_remove_empty ... ok
test compact_remove_remapped_keys ... ok
test delay_queue_poll_expired_when_empty ... ok
test expire_first_key_when_reset_to_expire_earlier ... ok
test expire_second_key_when_reset_to_expire_earlier ... ok
test expires_before_last_insert ... ok
test insert_after_ready_poll ... ok
test insert_before_first_after_poll ... ok
test insert_in_past_after_poll_fires_immediately ... ok
test insert_in_past_fires_immediately ... ok
test multi_delay_at_start ... ok
test multi_immediate_delays ... ok
test multi_reset ... ok
test remove_after_compact ... ignored, FIXME: Does not seem to work with WASI
test remove_after_compact_poll ... ignored, FIXME: Does not seem to work with WASI
test remove_at_timer_wheel_threshold ... ok
test remove_entry ... ok
test remove_expired_item ... ok
Error: failed to run main module `/home/runner/work/tokio/tokio/target/wasm32-wasi/debug/deps/time_delay_queue-f0c21ef322a44873.wasm`

Caused by:
    0: failed to invoke command default
    1: error while executing at wasm backtrace:
           0: 0x14360f - <unknown>!__rust_start_panic
           1: 0x142f1d - <unknown>!rust_panic
           2: 0x142ee4 - <unknown>!std::panicking::rust_panic_with_hook::hd6f3df478dab5dc5
           3: 0x1419aa - <unknown>!std::panicking::begin_panic_handler::{{closure}}::h7be61ca999d17a10
           4: 0x1418d4 - <unknown>!std::sys_common::backtrace::__rust_end_short_backtrace::h1c7217a99e6903e8
           5: 0x142552 - <unknown>!rust_begin_unwind
           6: 0x149a0d - <unknown>!core::panicking::panic_fmt::h86214d64a55c1d57
           7: 0x15016d - <unknown>!core::panicking::panic_display::h58fb45e5567db9b3
           8: 0x14fd1d - <unknown>!core::panicking::panic_str::hef6f4fea2c6da56a
           9: 0x14fceb - <unknown>!core::option::expect_failed::he9c7eebda4091106
          10: 0x140db4 - <unknown>!<std::time::Instant as core::ops::arith::Sub<core::time::Duration>>::sub::h9d1ea7149970c4f4
          11: 0xb656b - <unknown>!<tokio::time::instant::Instant as core::ops::arith::Sub<core::time::Duration>>::sub::h8709be5f2824f4d8
          12: 0x1423c - <unknown>!time_delay_queue::repeatedly_reset_entry_inserted_as_expired::{{closure}}::h5113477f19ab55d3
          13: 0x3e836 - <unknown>!<core::pin::Pin<P> as core::future::future::Future>::poll::h4fa8cd4e55f21ac9
          14: 0x3e8ce - <unknown>!<core::pin::Pin<P> as core::future::future::Future>::poll::h642aec97e6dd8367
          15: 0x3c5c7 - <unknown>!tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}::h1e04fe847e0eed21
          16: 0x3b926 - <unknown>!tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::hda5894a20edb5dea
          17: 0x3b74d - <unknown>!tokio::runtime::scheduler::current_thread::Context::enter::h0551c0eeeb75cb0b
          18: 0x3bfd4 - <unknown>!tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::h04e2d857a61424cc
          19: 0x3ad99 - <unknown>!tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}::hefeb9ef4b490f0b0
          20: 0x3aceb - <unknown>!tokio::macros::scoped_tls::ScopedKey<T>::set::h768b6b08d5c1ff74
          21: 0x3bd7c - <unknown>!tokio::runtime::scheduler::current_thread::CoreGuard::enter::h3549fc9053a672ee
          22: 0x3b4a4 - <unknown>!tokio::runtime::scheduler::current_thread::CoreGuard::block_on::h29331f6bd3fa002f
          23: 0x3b0df - <unknown>!tokio::runtime::scheduler::current_thread::CurrentThread::block_on::h252f5d42c4325f40
          24: 0x3aeb2 - <unknown>!tokio::runtime::runtime::Runtime::block_on::hd567519e6c4ccdf2
          25: 0x14000 - <unknown>!time_delay_queue::repeatedly_reset_entry_inserted_as_expired::h2e8d9e0d39e801f8
          26: 0x13e64 - <unknown>!time_delay_queue::repeatedly_reset_entry_inserted_as_expired::{{closure}}::hba3d5e0e60e4e59f
          27: 0x40eae - <unknown>!core::ops::function::FnOnce::call_once::h21770440c4a77129
          28: 0x78844 - <unknown>!test::__rust_begin_short_backtrace::h2f7251cabac8cc17
          29: 0x78824 - <unknown>!core::ops::function::FnOnce::call_once{{vtable.shim}}::h98ae03618d6ef851
          30: 0xa0133 - <unknown>!test::run_test::run_test_inner::hcac225d7d620b63b
          31: 0x8b209 - <unknown>!test::run_test::h2db965a4129a3fb3
          32: 0x871c7 - <unknown>!test::console::run_tests_console::h58c4d5717b25d8bf
          33: 0x9f467 - <unknown>!test::test_main::hc2411b2d7bc2431a
          34: 0x9f6b3 - <unknown>!test::test_main_static::hd744aa7211882b13
          35: 0x2c72f - <unknown>!time_delay_queue::main::h41b7fb16cef306a9
          36: 0x4107d - <unknown>!core::ops::function::FnOnce::call_once::h3c9e1736a43e028c
          37: 0x6f4ba - <unknown>!std::sys_common::backtrace::__rust_begin_short_backtrace::h47e6361a0d37dfca
          38: 0x3a287 - <unknown>!std::rt::lang_start::{{closure}}::h20bb17588a43137f
          39: 0x13a6bf - <unknown>!std::rt::lang_start_internal::h83cbd77d07ad3a3a
          40: 0x3a224 - <unknown>!std::rt::lang_start::h6ec71f408c08c5b2
          41: 0x2c753 - <unknown>!__main_void
          42: 0x1bfa - <unknown>!_start
       note: using the `WASMTIME_BACKTRACE_DETAILS=1` environment variable may show more debugging information
    2: wasm trap: wasm `unreachable` instruction executed

Stack backtrace:
   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
   1: wasmtime::trap::from_runtime_box
   2: wasmtime::func::Func::call_impl
   3: wasmtime::linker::Linker<T>::module::{{closure}}::{{closure}}
   4: wasmtime::func::Func::invoke
   5: wasmtime::func::Caller<T>::with
   6: wasmtime::trampoline::func::stub_fn
   7: <unknown>
   8: <unknown>
   9: wasmtime_setjmp
  10: wasmtime_runtime::traphandlers::<impl wasmtime_runtime::traphandlers::call_thread_state::CallThreadState>::with
  11: wasmtime_runtime::traphandlers::catch_traps
  12: wasmtime::func::Func::call_impl
  13: wasmtime_cli::commands::run::RunCommand::invoke_func
  14: wasmtime_cli::commands::run::RunCommand::load_main_module
  15: wasmtime_cli::commands::run::RunCommand::execute
  16: wasmtime::main
  17: std::sys_common::backtrace::__rust_begin_short_backtrace
  18: std::rt::lang_start::{{closure}}
  19: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/001a77fac33f6560ff361ff38f661ff5f1c6bf85/library/core/src/ops/function.rs:287:13
  20: std::panicking::try::do_call
             at /rustc/001a77fac33f6560ff361ff38f661ff5f1c6bf85/library/std/src/panicking.rs:483:40
  21: std::panicking::try
             at /rustc/001a77fac33f6560ff361ff38f661ff5f1c6bf85/library/std/src/panicking.rs:447:19
  22: std::panic::catch_unwind
             at /rustc/001a77fac33f6560ff361ff38f661ff5f1c6bf85/library/std/src/panic.rs:140:14
  23: std::rt::lang_start_internal::{{closure}}
             at /rustc/001a77fac33f6560ff361ff38f661ff5f1c6bf85/library/std/src/rt.rs:148:48
  24: std::panicking::try::do_call
             at /rustc/001a77fac33f6560ff361ff38f661ff5f1c6bf85/library/std/src/panicking.rs:483:40
  25: std::panicking::try
             at /rustc/001a77fac33f6560ff361ff38f661ff5f1c6bf85/library/std/src/panicking.rs:447:19
  26: std::panic::catch_unwind
             at /rustc/001a77fac33f6560ff361ff38f661ff5f1c6bf85/library/std/src/panic.rs:140:14
  27: std::rt::lang_start_internal
             at /rustc/001a77fac33f6560ff361ff38f661ff5f1c6bf85/library/std/src/rt.rs:148:20
  28: main
  29: <unknown>
  30: __libc_start_main
  31: <unknown>
error: test failed, to rerun pass `-p tokio-util --test time_delay_queue`

</details>

Darksonn commented:
Would it be possible to add a test for this?

satakuma commented on tokio/src/io/async_fd.rs:
```diff
@@ -508,6 +508,32 @@ impl<T: AsRawFd> AsyncFd<T> {
+    /// provides shared access to the inner IO resource when handling the
+    /// [`AsyncFdReadyGuard`].
+    #[allow(clippy::needless_lifetimes)] // The lifetime improves rustdoc rendering.
+    #[cfg(any(target_os = "linux", target_os = "android"))]
```
We should add the information about supported targets to the docs, please add this attribute here and in other places as well.
```suggestion
    #[cfg(any(target_os = "linux", target_os = "android"))]
    #[cfg_attr(docsrs, doc(cfg(any(target_os = "linux", target_os = "android"))))]
```

satakuma commented on tokio/src/io/interest.rs:
```diff
@@ -44,6 +44,10 @@ impl Interest {
 
+    /// Returns a `Interest` set representing priority completion interests.
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    pub const PRIORITY: Interest = Interest(mio::Interest::PRIORITY);
```
Most types from `tokio::net` have `ready()` and `try_io()` methods which take `Interest` as an argument, we should make sure they are well-behaved with this new type of interest. I think you will have to make some extra modifications to `src/io/ready.rs` and/or `src/io/interest.rs` to make those methods work.

folkertdev replied:
I think I've got them all. The one place I'm not entirely sure about is in `tokio/src/runtime/io/mod.rs`

```rust
impl Direction {
    pub(super) fn mask(self) -> Ready {
        match self {
            Direction::Read => Ready::READABLE | Ready::READ_CLOSED,
            Direction::Write => Ready::WRITABLE | Ready::WRITE_CLOSED,
        }
    }
}
```

I left this untouched, because it really seems to be about reading and writing here.

satakuma replied:
Yes, that's for AsyncRead/AsyncWrite only, thanks.

satakuma reviewed (commented):
> I'm not sure how/if this should be tested in tokio itself

Perhaps you take a look how [mio tests POLLPRI](https://github.com/tokio-rs/mio/blob/9bd0e9554d7d2dbb72d17fd3587e56e2796dd83f/tests/tcp_stream.rs#L801-L848) and add a similar test using types from Tokio.

folkertdev commented:
I'm happy to, this was a fun journey into the depths of tokio and mio (something I'd wanted to do for a while). 

I think I added `PRIORITY` in all the relevant places now. I added a test case for `AsyncFd` and for `TcpStream`, because for those I actually know how to trigger `EPOLLPRI`. Information on this is kind of scarce. I was under the impression that we could also listen for `EPOLLPRI` for a UDP socket to receive timestamping information, but I did not get that to work yet.

satakuma commented on tokio/src/io/interest.rs:
```diff
@@ -104,6 +128,8 @@ impl Interest {
             Interest::READABLE => Ready::READABLE | Ready::READ_CLOSED,
             Interest::WRITABLE => Ready::WRITABLE | Ready::WRITE_CLOSED,
+            #[cfg(any(target_os = "linux", target_os = "android"))]
+            Interest::PRIORITY => Ready::PRIORITY,
```
I don't think this branch is covered by the tests right now, it's only used by something like `try_io` from `TcpStream`. It would be great to have a test using `TcpStream::try_io` with the priority interest, for example a test which actually receives the OOB data inside the closure for `try_io`. You can use [`socket2::SockRef::recv_out_of_band`](https://docs.rs/socket2/latest/socket2/struct.SockRef.html#method.recv_out_of_band) for that. Note that you can receive only one byte with that method.

folkertdev replied:
Nice, I made the TCP test hit that branch.

poeschel replied:
I think, the `PRIORITY` branch should also cover `Ready::READ_CLOSED`, otherwise an `AsyncFd` with only `Interest::PRIORITY` will not get any notification if the fd closes and maybe wait forever? I don't know.

folkertdev replied:
that reasoning makes sense to me, but I don't actually know what the correct choice is here.

satakuma reviewed (approved):
Looks good to me, but I would like a +1 from @Darksonn as well.

poeschel commented on tokio/src/io/poll_evented.rs:
```diff
@@ -83,7 +91,7 @@ impl<E: Source> PollEvented<E> {
     #[cfg_attr(feature = "signal", allow(unused))]
     pub(crate) fn new(io: E) -> io::Result<Self> {
-        PollEvented::new_with_interest(io, Interest::READABLE | Interest::WRITABLE)
+        PollEvented::new_with_interest(io, ALL_INTEREST)
```
I would not change this, as this changes existing behaviour and I think `Interest::READABLE | Interest::WRITABLE` is a sane default most users would want.

folkertdev replied:
makes sense. I added a line to the docs specifying that only read and write interests are configured

poeschel commented on tokio/src/io/async_fd.rs:
```diff
@@ -508,6 +514,34 @@ impl<T: AsRawFd> AsyncFd<T> {
+    #[allow(clippy::needless_lifetimes)] // The lifetime improves rustdoc rendering.
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    #[cfg_attr(docsrs, doc(cfg(any(target_os = "linux", target_os = "android"))))]
+    pub async fn priority<'a>(&'a self) -> io::Result<AsyncFdReadyGuard<'a, T>> {
```
I think `priority` is not the best name here. The ones for the other interests are `readable` and `writeable`. Maybe `priority_ready` is better?

folkertdev replied:
yes, I like that much better

poeschel commented on tokio/src/io/async_fd.rs:
```diff
@@ -508,6 +514,34 @@ impl<T: AsRawFd> AsyncFd<T> {
+    #[allow(clippy::needless_lifetimes)] // The lifetime improves rustdoc rendering.
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    #[cfg_attr(docsrs, doc(cfg(any(target_os = "linux", target_os = "android"))))]
+    pub async fn priority_mut<'a>(&'a mut self) -> io::Result<AsyncFdReadyMutGuard<'a, T>> {
```
Same as above.

folkertdev replied:
fixed

poeschel reviewed (commented):
I have done something similar to your work here, but mine is not as "production ready" as your work. It is very well done. Thank you.
You can have a look at my code here: https://github.com/poeschel/tokio/tree/interest_prio
One thing I did differently is, I added direct priority support to the tokio `TcpStream`. This is maybe worth pulling in here. Priority support for the `TcpStream` also was in mio version 0.8.6.

And the other thing I did differently is I added a new waker for priority. (which then implies to also add the `Direction::Prioriry` (which sounds weird)). I thought about this and I am also not shure if it is right. But I think it is. If you have some events in the ready queue and then a priority event arrives, you want this to be handled indepently of the read events, right? This is what "priority" is for, isn't it?

poeschel commented:
> I think I added `PRIORITY` in all the relevant places now. I added a test case for `AsyncFd` and for `TcpStream`, because for those I actually know how to trigger `EPOLLPRI`. Information on this is kind of scarce. I was under the impression that we could also listen for `EPOLLPRI` for a UDP socket to receive timestamping information, but I did not get that to work yet.

Regarding a test for `AsyncFd` with `EPOLLPRI` events:
Linux gpio sysfs is able to fire `EPOLLPRI` events, when a gpio is configured to fire interrupts. You do not need real hardware for this, because there is a `gpio_mockup` driver in Linux. But the whole thing is a bit hard to setup, because most distributions do not have this driver activated, so you have to recompile your kernel and to setup and use it, you have to be a priviledged user.
Is it worth adding a test with these requirements?

poeschel commented:
> > Is it worth adding a test with these requirements?
> 
> Eh, maybe not. We wouldn't be able to test that in CI. However, I would like to know whether anyone has tried running the code and seen it work at least once for a sanity check.

I can test my use-case and the unit tests I wrote for my fork on this code tomorrow.

folkertdev commented:
thanks for your input @poeschel! 

re. pulling in more stuff: I'm open to it, but also think a separate later PR would be fine.

poeschel commented:
> > > Is it worth adding a test with these requirements?
> > 
> > 
> > Eh, maybe not. We wouldn't be able to test that in CI. However, I would like to know whether anyone has tried running the code and seen it work at least once for a sanity check.
> 
> I can test my use-case and the unit tests I wrote for my fork on this code tomorrow.

As promised, I tested today with my unit test and gpio sysfs in linux and my real world use-case with video4linux and both do work in principle. But I think we really should have a seperate worker for priority events:
1.  The tcp stream use-case: The linux tcp stack supports sending things over a tcp connection with MSG_OOB set. On the receiving side this is then received over the same tcp connection, but is priorized and with POLLPRI set. This is an urgent message, that will arrive ahead of the normal buffered tcp stream actual in transit. By reusing the read waker for priority events in tokio, the priority mechanism will not work anymore. You will receive the message, but later than possible if we had another independent waker.
2. The video4linux use case: [Linux does support POLLPRI events on certain video4linux devices.](https://linuxtv.org/downloads/v4l-dvb-apis/userspace-api/v4l/dev-event.html)  There are different events to receive, like frame sync, resolution changed or motion detected or something like that. That means POLLPRI is used like a second channel on the same video device. You receive your video pixel data by `read`ing from the video device and you receive interesing events over POLLPRI. These should not be intermixed. This will not work very well with tokio when we reuse the read waker for priority events.

I think it is better to add a seperate priority waker, or am I missing something?

Darksonn commented:
Adding a separate waker for priority events makes sense to me.

folkertdev commented:
@poeschel I've been looking at this. One thing I don't see is how the current example would use this new priority direction/waker. You seem to be using it in a custom `Stream` implementation.

what are the consequences of adding the custom waker for the functionality in this PR so far? or is it really a primitive that is needed for e.g. TCP?

folkertdev commented:
here's what I understand:

Because we don't have async traits, traits that deal with async often have functions that return `Poll<T>`. @poeschel  's [branch](https://github.com/tokio-rs/tokio/compare/master...poeschel:tokio:interest_prio) has an example using `futures::stream::Stream`, but I think tokio's `AsyncRead` and `AsyncWrite` are similar.

```rust
impl Stream for AsyncV4l2EventHandle {
    type Item = io::Result<i32>;

    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
        loop {
            let blah = self.asyncfd.poll_priority_ready_mut(cx);
            let mut guard = ready!(blah)?;
            let res = guard.try_io(|inner| inner.get_mut().get_event());
            guard.clear_ready();
            match res {
                Err(TryIoError { .. }) => {
                    continue;
                },
                Ok(Ok(event)) => {
                    return Poll::Ready(Some(Ok(event)));
                },
                Ok(Err(err)) => {
                    return Poll::Ready(Some(Err(err.into())));
                }
            }
        }
    }
}
```

The `poll_priority_ready_mut` function is analogous to `poll_read_ready`. I believe the `Stream` implementation above runs into exactly the reason that these `poll_` functions exist, from their docs: 

> This method is intended for cases where creating and pinning a future
> via [`readable`] is not feasible. Where possible, using [`readable`] is
> preferred, as this supports polling from multiple tasks at once.

There is nowhere to pin a future in that impl (right?)

<hr>

So assuming that we want `poll_priority_ready_mut`, it needs to be implemented. This is how `poll_read_ready_mut` works

```rust
impl AsyncFd<T> {
    pub fn poll_read_ready_mut<'a>(
        &'a mut self,
        cx: &mut Context<'_>,
    ) -> Poll<io::Result<AsyncFdReadyMutGuard<'a, T>>> {
        let event = ready!(self.registration.poll_read_ready(cx))?;

        Ok(AsyncFdReadyMutGuard {
            async_fd: self,
            event: Some(event),
        })
        .into()
    }
}

...

impl Registration {
    // Uses the poll path, requiring the caller to ensure mutual exclusion for
    // correctness. Only the last task to call this function is notified.
    pub(crate) fn poll_read_ready(&self, cx: &mut Context<'_>) -> Poll<io::Result<ReadyEvent>> {
        self.poll_ready(cx, Direction::Read)
    }
}
```

so working by analogy, we need some `Direction::Priority` to pass to `poll_ready`. and that ultimately needs an extra waker for priority events.

<hr>

The above makes sense to me, and the conclusion seems to be that for correct behavior, an extra waker for priority events is needed. But maybe I'm missing something here?

carllerche commented:
To avoid the waker issue, I would not expose priority via a "poll" method. The async fns do not need a waker per readiness type. You can always convert a future returned by an `async fn` to a nameable future by doing `Box<dyn Future<...>>`.

carllerche commented:
It seems to me the key issue here is `AsyncFd` does not have an `async fn ready` like `TcpStream` does ([here](https://docs.rs/tokio/latest/tokio/net/struct.TcpStream.html#method.ready)). Let's start with a PR that adds that.

Then, we can add "priority" to `Interest,` and the way to listen for priority events is to call `fd.ready(Interest::PRIORITY).await` (or whatever).

folkertdev commented:
this PR already adds 

    pub async fn priority_ready<'a>(&'a self) -> io::Result<AsyncFdReadyGuard<'a, T>> {

which is follows the pattern of

    pub async fn readable<'a>(&'a self) -> io::Result<AsyncFdReadyGuard<'a, T>> {

and then adds a `PRIORITY` variant to `Interest`. All of that works fine. 

<hr>

implementing `ready` looks like a nice addition, but I think you run into issues when an IO operation blocks. At that point you need to clear the readiness and with just a `Ready` value that is not possible.

The `AsyncFdReadyGuard` returned by `readable` and friends has only an `Option<ReadyEvent>`, and therefore cannot always give you a `Ready` value (given the current implementation)

<hr>

I'm not sure how boxing helps in the example above. The problem is that there is no way to hold on to a future between calls to `poll_next`. The code below would compile

```
fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
    loop {
        let fut = self.asyncfd.priority_ready();
        let mut guard = ready!(fut)?;

        ...
    }
}
```

but the call to `priority_ready` re-creates the future every time. I'm assuming that has performance downsides, but maybe it's fine.

<hr>

all in all, I'd suggest that we declare polling out-of-scope for this PR, and focus just on the two functions from the title, `priority_ready` and `priority_ready_mut`.

folkertdev commented:
is this PR blocked on something I can help with?

I think the CI failure is a flake. We've had a bunch of discussion here, which has led to just limiting the scope of this PR and leaving support for various `poll_` functions for the future. 

if it's blocked on someone having the time/energy to look at this that's totally fine, I'd just like to know what the status is.

Darksonn commented:
My immediate reaction is that I would prefer to add a `ready` method that takes an `Interest`, rather than adding a `priority_ready` method.

folkertdev commented:
I'm not sure that `fn ready(&self, interest: Interest) -> Result<Ready>` works as an API here, at least by itself, because the readiness cannot be cleared. Clearing of the readiness is taken care of by tokio for its builtin types (like `tokio::net::TcpStream`), but as far as I understand must be performed manually when using the standard library types.

An alternative is to make the exising `readiness` function public, and rename it to `ready` (similarly for the `_mut` version)

```rust
pub async fn ready(&self, interest: Interest) -> io::Result<AsyncFdReadyGuard<'_, T>> {
    let event = self.registration.readiness(interest).await?;

    Ok(AsyncFdReadyGuard {
        async_fd: self,
        event: Some(event),
    })
}
```

This provides a guard, which can be used to clear the readiness. However, the guard does not always contain a `Ready` value, because it is defined as

```rust
pub struct AsyncFdReadyGuard<'a, T: AsRawFd> {
    async_fd: &'a AsyncFd<T>,
    event: Option<ReadyEvent>,
}
```

I _think_ we can work around that by providing 

```rust
impl<'a, Inner: AsRawFd> AsyncFdReadyGuard<'a, Inner> {
    /// Get the [`Ready`] value associated with this guard.
    ///
    /// [`Ready`]: super::Ready
    pub fn ready(&self) -> Ready {
        match &self.event {
            Some(event) => event.ready,
            None => Ready::EMPTY,
        }
    }

    ...
}
```

which would then allow us to write the example from `tokio::net::TcpStream::ready` as follows

```rust
use std::error::Error;
use std::io;
use std::io::{Read, Write};
use std::net::TcpStream;
use tokio::io::unix::AsyncFd;
use tokio::io::Interest;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let stream = TcpStream::connect("127.0.0.1:8080"?;
    stream.set_nonblocking(true)?;
    let stream = AsyncFd::new(stream)?;

    loop {
        let mut guard = stream
            .ready(Interest::READABLE | Interest::WRITABLE)
            .await?;

        if guard.ready().is_readable() {
            let mut data = vec![0; 1024];
            // Try to read data, this may still fail with `WouldBlock`
            // if the readiness event is a false positive.
            match stream.get_ref().read(&mut data) {
                Ok(n) => {
                    println!("read {} bytes", n);
                }
                Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
                    guard.clear_ready();
                    continue;
                }
                Err(e) => {
                    return Err(e.into());
                }
            }
        }

        if guard.ready().is_writable() {
            // Try to write data, this may still fail with `WouldBlock`
            // if the readiness event is a false positive.
            match stream.get_ref().write(b"hello world") {
                Ok(n) => {
                    println!("write {} bytes", n);
                }
                Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
                    guard.clear_ready();
                    continue;
                }
                Err(e) => {
                    return Err(e.into());
                }
            }
        }
    }
}
```

This behaves like the original example for tokio `TcpStream`, at least I believe it does. One little doubt I have is whether the clearing of the readiness is actually correct here, because a blocking read would also clear the readiness of the write and vice versa. I'm not really sure whether that is a problem. 

So, thoughts on this? does this look like a more promising direction? It certainly does cut down on the amount of changes required to support more `Interest` variants.

folkertdev commented:
I made the changes from the above comment. Generally works out well. 

I also had to remove a test that used tokio's `TcpStream`, because it will not register with the `PRIORITY` interest by default. It seems like a good idea to me to keep such changes outside of this PR, TCP needs extra work anyway around poling.

Darksonn commented on tokio/src/io/async_fd.rs:
```diff
@@ -440,7 +442,82 @@ impl<T: AsRawFd> AsyncFd<T> {
+    ///                 Ok(n) => {
+    ///                     println!("read {} bytes", n);
+    ///                 }
+    ///                 Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+    ///                     guard.clear_ready();
+    ///                     continue;
+    ///                 }
```
This seems problematic. We haven't attempted a write, so if it is also ready for writing, then we incorrectly clear the write readiness.

Darksonn replied:
I'm not sure how we can best approach this ... we will probably need some way to clear individual parts of the readiness.

folkertdev replied:
yes, that's a consequence of exposing `AsyncFd::ready`. Something like this might work on the guards

```rust
    /// Indicates to tokio that the file descriptor no longer has a specific readiness.
    /// The internal readiness flag will be cleared, and tokio will wait for the
    /// next edge-triggered readiness notification from the OS.
    ///
    /// This function is useful in combination with the [`AsyncFd::ready`] method when a
    /// combined interest like `Interest::READABLE | Interest::WRITABLE` is used.
    ///
    /// It is critical that this function not be called unless your code
    /// _actually observes_ that the file descriptor is _not_ ready for the provided `Ready`.
    /// Do not call it simply because, for example, a read succeeded; it should be called
    /// when a read is observed to block. Only clear the specific readiness that is observed to
    /// block. For example when a read blocks when using a combined interest,
    /// only clear `Ready::READABLE`.
    pub fn clear_ready_exact(&mut self, ready: Ready) {
        if let Some(event) = &mut self.event {
            self.async_fd
                .registration
                .clear_readiness(event.with_ready(ready));

            // the event is no longer ready for the readiness that was just cleared
            event.ready = event.ready - ready;
        }
    }
```

based on the implementation I think the behavior is correct

```rust
    pub(crate) fn clear_readiness(&self, event: ReadyEvent) {
        // This consumes the current readiness state **except** for closed
        // states. Closed states are excluded because they are final states.
        let mask_no_closed = event.ready - Ready::READ_CLOSED - Ready::WRITE_CLOSED;

        // result isn't important
        let _ = self.set_readiness(None, Tick::Clear(event.tick), |curr| curr - mask_no_closed);
    }
```

my only doubt is whether clearing the readiness for the same tick multiple times works as intented. I think it does based on the implementation of `set_readiness`.

Darksonn commented on tokio/src/io/async_fd.rs:
```diff
@@ -440,7 +442,86 @@ impl<T: AsRawFd> AsyncFd<T> {
     }
 
-    async fn readiness(&self, interest: Interest) -> io::Result<AsyncFdReadyGuard<'_, T>> {
+    /// Waits for any of the requested ready states, returning a
+    /// [`AsyncFdReadyGuard`] that must be dropped to resume
+    /// polling for the requested ready states.
+    ///
+    /// The function may complete without the file descriptor being ready. This is a
+    /// false-positive and attempting an operation will return with
+    /// `io::ErrorKind::WouldBlock`. The function can also return with an empty
+    /// [`Ready`] set, so you should always check the returned value and possibly
+    /// wait again if the requested states are not set.
+    ///
+    /// This method takes `&self`, so it is possible to call this method
+    /// concurrently with other methods on this struct. This method only
+    /// provides shared access to the inner IO resource when handling the
+    /// [`AsyncFdReadyGuard`].
```
Your changes seem reasonable to me. I'd like to also call `clear_read_exact` out prominently in the main text of the documentation for `ready`.

Darksonn reviewed (commented):
A `clear_ready_exact` method makes sense to me.

carllerche commented on tokio/src/io/async_fd.rs:
```diff
@@ -641,22 +753,47 @@ impl<T: AsRawFd> Drop for AsyncFd<T> {
+    /// when a read is observed to block. Only clear the specific readiness that is observed to
+    /// block. For example when a read blocks when using a combined interest,
+    /// only clear `Ready::READABLE`.
+    pub fn clear_ready_exact(&mut self, ready: Ready) {
```
It looks like this fn doesn't check if the supplied `ready` is a superset of the event's readiness. If it is, it seems that `self.event` should be cleared (as in `clear_ready()`).

* Could you add a test for this method covering these cases?
* Could you include a doc example as well?

I don't *love* the name `clear_ready_exact`,  but I also don't have a better idea, so I'm OK sticking with it. I do like that it starts with the `clear_ready` prefix, though.

I asked ChatGPT for some ideas, I kind of like `clear_ready_matching` or `clear_ready_with`.

ChatGPT's answer:

> Choosing a method name can often be a matter of personal style and the idiomatic conventions in a codebase, but here are some suggestions for the method name:
> 
> 1. `clear_ready_intersection`: This name clearly communicates that the method clears the intersection of readiness.
> 
> 2. `clear_ready_matching`: This implies that only the readiness bits that match those of the argument will be cleared.
> 
> 3. `clear_ready_by`: This could suggest that the readiness is cleared "by" the criteria of the given `Ready` instance.
> 
> 4. `clear_ready_with`: The "with" here could suggest that the clearing action is done in accordance with the `Ready` instance that's passed.
> 
> 5. `clear_ready_bits`: This is more general, but the "bits" could indicate that you're passing another bitfield to determine what's cleared.
> 
> 6. `clear_ready_shared`: This name would indicate that it's clearing readiness that's shared between the socket and the `Ready` instance.
> 
> 7. `clear_ready_conditional`: This might suggest that there's some condition (the `Ready` argument) that determines which readiness is cleared.
> 
> 8. `clear_ready_overlap`: The word "overlap" denotes the intersection concept.
> 
> Remember, it's also important to document the behavior of this method clearly in comments or other forms of documentation so that future developers understand how it works.

folkertdev replied:
> I don't love the name clear_ready_exact, but I also don't have a better idea

fully agree! in context, I think I like `clear_ready_matching` best:

    ///                 Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
    ///                     guard.clear_ready_exact(Ready::READABLE);
    ///                     continue;
    ///                 }

    ///                 Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
    ///                     guard.clear_ready_matching(Ready::READABLE);
    ///                     continue;
    ///                 }

    ///                 Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
    ///                     guard.clear_ready_with(Ready::READABLE);
    ///                     continue;
    ///                 }

For testing, I added some basic tests. What is not (yet) tested is that the event is fully cleared. This is tricky, because clearing the event is not observable from the outside. The same is actually true for the existing `clear_ready` function. I'm also not sure it's that relevant. it's an optimization, but not required for correctness.

It is hard to come up with a succinct example here. This function only makes sense with a combined interest, or when using the priority interest but it is linux-only. So, I can't really come up with anything better than just copy-pasting the example from `ready`. I'm happy to do that, but it would duplicate that example 4 times in total (ready, ready_mut, and clear_ready_matching for the two guard types. There is risk of introducing inconsistencies. Anyway, thoughts?

folkertdev replied:
the `clear_ready_matching` functions now have the example from their matching `ready` function as an example.

carllerche commented on tokio/src/io/async_fd.rs:
```diff
@@ -449,7 +540,28 @@ impl<T: AsRawFd> AsyncFd<T> {
+    ///
+    /// This method takes `&mut self`, so it is possible to access the inner IO
+    /// resource mutably when handling the [`AsyncFdReadyMutGuard`].
+    pub async fn ready_mut(
```
A doc example would be nice to have here. You could copy the one above and use some sort of mutation.

folkertdev replied:
based on some github searches, it looks like the `_mut` functions are hardly used at all. The only uses I could find are around linux eBPF stuff. Even there it might work to implement traits on a reference, e.g. there is a `impl Write for &UdpSocket` so that you don't need to have a mutable reference to a UdpSocket but can still implement the `Write` trait, which does take a mutable reference.

On the other hand, the function deserves an example, and I couldn't think of any better examples, so I've just replicated the one from `ready` here.

carllerche replied:
A basic example is totally fine. Part of the reason is to just ensure we also have a test for the public API (via doc tests).

carllerche commented on tokio/src/io/async_fd.rs:
```diff
@@ -751,22 +902,47 @@ impl<'a, Inner: AsRawFd> AsyncFdReadyGuard<'a, Inner> {
+    /// when a read is observed to block. Only clear the specific readiness that is observed to
+    /// block. For example when a read blocks when using a combined interest,
+    /// only clear `Ready::READABLE`.
+    pub fn clear_ready_exact(&mut self, ready: Ready) {
```
Same comments as above apply here.

carllerche commented on tokio/src/io/ready.rs:
```diff
@@ -240,11 +282,25 @@ impl ops::Sub<Ready> for Ready {
-            .field("is_read_closed", &self.is_read_closed())
-            .field("is_write_closed", &self.is_write_closed())
-            .finish()
+        #[cfg(any(target_os = "linux", target_os = "android"))]
```
You shouldn't need to dup the full block. You can start the `fmt` with:

```rust
let fmt = fmt.debug_struct("Ready");`
```

then have all the fields that are common across platforms.

Then:

```rust
#[cfg(any(target_os = "linux", target_os = "android"))]
fmt.field("is_priority", &self.is_priority_ready())
```

folkertdev replied:
ah nice. I tried this before in a slightly different way and ran into lifetime issues.

carllerche commented on tokio/src/io/ready.rs:
````diff
@@ -144,6 +163,23 @@ impl Ready {
+    /// ```
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    #[cfg_attr(docsrs, doc(cfg(any(target_os = "linux", target_os = "android"))))]
+    pub fn is_priority_ready(self) -> bool {
````
This should probably just be `is_priority`. Being on the `Ready` struct already implies ready.

carllerche reviewed (changes requested):
Looks good. At a high level, the direction is solid. I left some comments inline requesting changes, but those are more implementation details and nothing fundamental.

Thanks for the work :+1:.

folkertdev commented:
wasm/wasi CI fails with

```
error[E0433]: failed to resolve: use of undeclared crate or module `log`
  --> /home/runner/.cargo/registry/src/index.crates.io-6f17d[22](https://github.com/tokio-rs/tokio/actions/runs/5114795867/jobs/9195428801?pr=5566#step:10:23)bba15001f/mio-0.8.7/src/sys/wasi/mod.rs:82:13
   |
82 |             log::warn!(
   |             ^^^ use of undeclared crate or module `log`
```

which is weird. In the `Cargo.toml` I specify we now use version `0.8.6` of mio, but the `Cargo.lock` file does actually show `0.8.7`. Maybe something with feature flags changed in that version of mio?

Darksonn commented:
The mio issue has been fixed upstream. I reran ci and it should pass now.

carllerche reviewed (approved):
LGTM, thanks

Darksonn commented on tokio/Cargo.toml:
```diff
@@ -103,7 +103,7 @@ pin-project-lite = "0.2.0"
 # Everything else is optional...
 bytes = { version = "1.0.0", optional = true }
-mio = { version = "0.8.4", optional = true, default-features = false }
+mio = { version = "0.8.6", optional = true, default-features = false  }
```
```suggestion
mio = { version = "0.8.6", optional = true, default-features = false }
```

Darksonn reviewed (approved):
One nit, otherwise LGTM.

folkertdev commented:
thanks for all the review/help, I really like where this ended up!

poeschel commented:
Also, thank you very much from my side for your work!
